Add Guard-Malloc and ASAN as modifiers to test expectations

rniwa opened the pull request:
#### 77cc784b7c9e760613ae2fe7849e3215184be868
<pre>
Add Guard-Malloc and ASAN as modifiers to test expectations
<a href="https://bugs.webkit.org/show_bug.cgi?id=250407">https://bugs.webkit.org/show_bug.cgi?id=250407</a>

Reviewed by David Kilzer and Jonathan Bedard.

Add the support for &quot;ASAN&quot; and &quot;Guard-Malloc&quot; as test expectation modifiers.
These are build types like &quot;Release&quot; and &quot;Debug&quot;.

Use that to skip crypto/crypto-random-values-oom.html on guard malloc bots.

* LayoutTests/platform/mac/TestExpectations:
* Tools/Scripts/webkitpy/layout_tests/models/test_configuration_unittest.py:
(make_mock_all_test_configurations_set):
(TestConfigurationConverterTest.test_to_config_set_with_asan_and_guard_malloc):
(TestConfigurationConverterTest.test_macro_expansion):
(TestConfigurationConverterTest.test_to_specifier_lists):
* Tools/Scripts/webkitpy/port/base.py:
(Port):
(Port.test_configuration):

Canonical link: <a href="https://commits.webkit.org/258760@main">https://commits.webkit.org/258760@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/39a86cd84e547a14d6eb2e17e9db52ccfa1f9e23

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102861 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11978 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35887 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112114 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/172336 "Hash 39a86cd8 for PR 8482 does not build (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106824 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12997 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2887 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95109 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109782 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108635 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/9979 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/93179 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37626 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/106374 "Passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91824 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24711 "Found 2 new test failures: http/tests/security/clean-origin-css-exposed-resource-timing.html, http/tests/security/cross-origin-clean-css-resource-timing.html (failure)") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/79360 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5431 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26125 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5579 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/2577 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11597 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45616 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7325 "Built successfully") | | | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3205 "Failed to push commit to Webkit repository") | | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Tools/Scripts/webkitpy/layout_tests/models/test_configuration_unittest.py:
```diff
@@ -244,6 +310,25 @@ def test_macro_expansion(self):
+            TestConfiguration('lucid', 'x86_64', 'release', 'asan'),
+            TestConfiguration('lucid', 'x86_64', 'release', 'guard-malloc'),
+        ])
+        self.assertEqual(converter.to_config_set(set(['x86_64'])), configs_to_match)
```
Would it make sense to have a test where `'asan'` or `'guard-malloc'` art port of the config set?

Maybe like this?  (This is just a guess at whether this code is correct.)

```
        configs_to_match = set([
             TestConfiguration('lucid', 'x86_64', 'debug', 'asan'),
             TestConfiguration('lucid', 'x86_64', 'release', 'asan'),
         ])
         self.assertEqual(converter.to_config_set(set(['x86_64', 'asan'])), configs_to_match)
```

rniwa replied:
These tests DO test that.

ddkilzer reviewed (approved):
r=me, but it would be good if you could get @JonWBedard to review as well as he likely knows this code better.

JonWBedard commented on Tools/Scripts/webkitpy/layout_tests/models/test_configuration.py:
```diff
@@ -31,15 +31,17 @@
 
 class TestConfiguration(object):
-    def __init__(self, version, architecture, build_type):
+    def __init__(self, version, architecture, build_type, build_style='normal'):
```
I think the consequences of adding `build_style` are more dramatic than what's reflected in this change. For one, we'd need to pipe that same information through to the results database as a new configuration column.

What we've done in other places (namely, the results database) is treat `ASAN` and `Guard-Malloc` as `build_type`s. For `ASAN`, that is, strictly speaking, true. It's not exactly true for `Guard-Malloc` (ie, you can have `release`+`Guard-Malloc` or `debug`+`Guard-Malloc`, but in practice, we only ever run `Guard-Malloc` one way, and it seems unlikely that we would ever need to differentiate between `debug` and `release` Guard-Malloc results. I think that would turn this change into a much smaller one, we'd just need to add `asan` and `guard-malloc` to `ALL_BUILD_TYPES`, I believe. Is there a reason we aren't solving this problem that way?

rniwa replied:
> I think the consequences of adding build_style are more dramatic than what's reflected in this change. For one, we'd need to pipe that same information through to the results database as a new configuration column.

That's already done in base.configuration_for_upload:
```python
    def configuration_for_upload(self, host=None):
        from webkitpy.results.upload import Upload

        configuration = self.test_configuration()
        host = self.host or host

        if self.get_option('guard_malloc'):
            style = 'guard-malloc'
        elif self._config.asan:
            style = 'asan'
        else:
            style = configuration.build_type

        return Upload.create_configuration(
            platform=host.platform.os_name,
            version=str(host.platform.os_version),
            version_name=host.platform.os_version_name(INTERNAL_TABLE) or host.platform.os_version_name(),
            architecture=configuration.architecture,
            style=style,
            sdk=host.platform.build_version(),
            flavor=self.get_option('result_report_flavor'),
            model=self.get_option('model'),
        )
```

JonWBedard reviewed (changes requested):
Is there a reason we aren't just treating `Guard-Malloc` and `ASAN` as `build_type`s? It's not, strictly speaking, correct for `Guard-Malloc`, but the consequence of that is just that we won't be able to differentiate between Release+Guard-Malloc and Debug+Guard-Malloc, and I can't see us wanting to do that.

rniwa committed:
Add Guard-Malloc and ASAN as modifiers to test expectations
https://bugs.webkit.org/show_bug.cgi?id=250407

Reviewed by David Kilzer and Jonathan Bedard.

Add the support for "ASAN" and "Guard-Malloc" as test expectation modifiers.
These are build types like "Release" and "Debug".

Use that to skip crypto/crypto-random-values-oom.html on guard malloc bots.

* LayoutTests/platform/mac/TestExpectations:
* Tools/Scripts/webkitpy/layout_tests/models/test_configuration_unittest.py:
(make_mock_all_test_configurations_set):
(TestConfigurationConverterTest.test_to_config_set_with_asan_and_guard_malloc):
(TestConfigurationConverterTest.test_macro_expansion):
(TestConfigurationConverterTest.test_to_specifier_lists):
* Tools/Scripts/webkitpy/port/base.py:
(Port):
(Port.test_configuration):

Canonical link: https://commits.webkit.org/258760@main
